File: tomboy_boot/src/main/java/com/fly/clstudy/global/exceptions/GlobalExceptionHandler.java
package com.fly.clstudy.global.exceptions;

import com.fly.clstudy.global.https.ReqData;
import com.fly.clstudy.global.https.RespData;
import com.fly.clstudy.global.jpa.util.EmpClass;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseBody;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.util.LinkedHashMap;
import java.util.Map;

@ControllerAdvice
@RequiredArgsConstructor
@Slf4j
public class GlobalExceptionHandler {

    private final ReqData rq;

    @ExceptionHandler(Exception.class)
    public ResponseEntity<Object> handleException(Exception ex) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("resultCode", "500-1");
        body.put("statusCode", 500);
        body.put("msg", ex.getLocalizedMessage());

        LinkedHashMap<String, Object> data = new LinkedHashMap<>();
        body.put("data", data);

        StringWriter sw = new StringWriter();
        PrintWriter pw = new PrintWriter(sw);
        ex.printStackTrace(pw);
        data.put("trace", sw.toString().replace("\t", "    ").split("\\r\\n"));

        String path = rq.getCurrentUrlPath();
        data.put("path", path);

        return new ResponseEntity<>(body, HttpStatus.INTERNAL_SERVER_ERROR);
    }


    @ExceptionHandler(GlobalException.class)
    @ResponseBody
    public ResponseEntity<RespData<EmpClass>> handleException(GlobalException ex) {

        RespData<EmpClass> rsData = ex.getRsData();

        return ResponseEntity.status(rsData.getStatusCode()).body(rsData);
    }

    @ExceptionHandler(MethodArgumentNotValidException.class)
    @ResponseBody
    public ResponseEntity<RespData<EmpClass>> handleException(MethodArgumentNotValidException ex) {
        String resultCode = "400-" + ex.getBindingResult().getFieldError().getCode();
        String msg = ex.getBindingResult().getFieldError().getField() + " : " + ex.getBindingResult().getFieldError().getDefaultMessage();

        return handleException(
                new GlobalException(
                        resultCode,
                        msg
                )
        );
    }
}

File: k_rstudy/src/main/kotlin/com/krstudy/kapi/global/exception/GlobalExceptionHandler.kt
package com.krstudy.kapi.com.krstudy.kapi.global.exception

import com.krstudy.kapi.global.exception.CustomException
import com.krstudy.kapi.global.exception.GlobalException
import com.krstudy.kapi.global.https.ReqData
import com.krstudy.kapi.global.https.RespData
import com.krstudy.kapi.standard.base.Empty
import lombok.RequiredArgsConstructor
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.ResponseStatus
import java.io.PrintWriter
import java.io.StringWriter

@ControllerAdvice
@RequiredArgsConstructor
class GlobalExceptionHandler (private val rq: ReqData){

    @ExceptionHandler(CustomException::class)
    fun handleCustomException(ex: CustomException): ResponseEntity<Map<String, Any>> {
        val errorResponse = mapOf(
            "code" to ex.errorCode.code,
            "message" to ex.errorCode.message
        )
        return ResponseEntity(errorResponse, HttpStatus.valueOf(ex.errorCode.code.split("-")[0].toInt()))
    }
    // 자연스럽게 발생시킨 예외처리
    private fun handleApiException(ex: Exception): ResponseEntity<Any> {
        val body = LinkedHashMap<String, Any>().apply {
            put("resultCode", "500-1")
            put("statusCode", 500)
            put("msg", ex.localizedMessage)

            val data = LinkedHashMap<String, Any>()
            put("data", data)

            val sw = StringWriter()
            val pw = PrintWriter(sw)
            ex.printStackTrace(pw)
            data["trace"] = sw.toString().replace("\t", "    ").split(Regex("\\r\\n"))

            val path = rq.getCurrentUrlPath()
            data["path"] = path

            put("success", false)
            put("fail", true)
        }

        return ResponseEntity(body, HttpStatus.INTERNAL_SERVER_ERROR)
    }

    // 개발자가 명시적으로 발생시킨 예외처리
    @ExceptionHandler(GlobalException::class)
    @ResponseStatus // 참고로 이 코드의 역할은 error 내용의 스키마를 타입스크립트화 하는데 있다.
    fun handle(ex: GlobalException): ResponseEntity<RespData<Empty>> {
        val status = HttpStatus.valueOf(ex.rsData.statusCode)
        rq.setStatusCode(ex.rsData.statusCode)

        return ResponseEntity(ex.rsData, status)
    }

    // 로깅 시 민감한 정보 마스킹
    private fun maskSensitiveInfo(info: String): String {
        return info.take(3) + "*".repeat(info.length - 3)
    }

}